### ISSUE TITLE
Why is the size of  rpb [2K-1,2K-1] inconsistent to kernel size [K,K]? 

### ISSUE THREAD

**Issue by ACuOoOoO** [SHI-Labs/NATTEN#165]
Thank you very much for your excellent work. I wonder rpb should be added to attention weight in shape of [K, K]. However, the shape of rpb in 2d NATTEN is [2K-1,2K-1].  It is hard for me to find the answer from the C codes. I hope Pytorch implementations can be offered for performance comparison and code understanding.

**Comment by alihassanijr** [comment]
Thank you for your interest. I'd recommend referring to [Swin Transformer](https://arxiv.org/abs/2103.14030) where RPB for local attention is described in detail (page 5, sec 3.2).

It boils down to this: attention window size being a certain size does not guarantee there are exactly that many relative positions. Relative positions are defined with respect to queries, and if all queries were to be the "center" of their attention window, you would be right, the number of positions relative to the query position would be exactly the size of the window.

**Comment by ACuOoOoO** [comment]
Thanks for your quick and detailed reply. RPB in SWin should be in shape of [2K-1,2K-1], since the queries are sampled from a attention window in shape of [k,k]. However, there is only one query centered at the window in NA, isn't there?

**Comment by alihassanijr** [comment]
No, sliding window attention guarantees the query being centered. Neighborhood attention breaks that in the corner cases so that queries attend to the same number of context tokens regardless of their position.

Neighborhood attention:

![na_corner](https://github.com/user-attachments/assets/138c5775-7cbb-4098-8d88-b2836f3954e9)

Sliding window attention:

![sliding_window_corner](https://github.com/user-attachments/assets/12acbce0-6eef-47d2-8983-6c5fb1aeac0e)

**Comment by ACuOoOoO** [comment]
Thanks very much. I have ignored the cases on corners and edges. Would the special operations for pixels on corners and edges bring additional computation cost that is not worth the improvement on performance, especially in the case of large size input, for example, 800x600? Can you add an option to enable naive 2d sliding window attention for higher efficiency?

**Comment by alihassanijr** [comment]
What do you mean by ignoring the corner cases?

And in theory, no I disagree. Unlike convolution, I think if your input size grows, so should your window size, otherwise as you grow your input size and keep your window size constant, you approach a linear projection instead of attention.

The number of corner cases grows with kernel size and dilation.

As far as efficiency goes, any performant implementation of either neighborhood or sliding window attention doesn't compute the operation query-by-query, and the only difference would be in how sub attention weights are masked, which will have little to no effect on speed observable by users, at least that's what I would guess.

But if performance and efficiency is a concern, I wouldn't even use RPB to begin with, because it is terribly inefficient. RPB makes even fused attention kernels memory bandwidth bound, and in those case you're typically already inefficient in math/tensor core utilization.

**Comment by ACuOoOoO** [comment]
It is good that efficiency is not affected.  Thank you for your patient and detailed response！

The relatively precise relative positional encoding is crucial for my study in theory. This relative positional encoding should be entirely independent of absolute positions. It means some existing alternatives to RPB, such as ROPE, might not be suitable. I will explore other solutions on my own. Thank you again!

**Comment by alihassanijr** [comment]
Of course; let me know if you have any other questions.

>The relatively precise relative positional encoding is crucial for my study in theory.

I see; that's totally fine. Though I generally wouldn't expect --fused-- attention kernels with RPB or fully materialized attention masks/biases to have good performance, especially at scale. That's why we dropped RPB support in our fused kernels (forward pass still supports it though, but not backward pass.)

In the unfused case and at scale I wouldn't care too much about efficiency compared to feasibility.

**Comment by ACuOoOoO** [comment]
I'm sorry to bother you again. Assume the output size of the na2d_qk function is [b, n, h, w, L], where L=KxK. For a query centered in the patch, can I obtain its attention score to its neighbors in patch by "unflattening" the last dim to [K,K]?

**Comment by alihassanijr** [comment]
Yes you can; the last mode/axis in attention maps produced by `na{1,2,3}d_qk` can be unrolled into `[K_h, K_w]`.

**Comment by alihassanijr** [comment]
The issue appears to be resolved, so closing now. Feel free to reopen if you still have questions.

**Comment by Temo233233** [comment]
Hi @alihassanijr  , sorry to interrupt. I still have a question regarding the computational efficiency of RPB.
You mentioned that calculating RPB in NAT is inefficient. However, if all the queries are centered , all attention windows will share the same RPB encoding.

```
    x = pad(x, (0, 0, self.pad_size, self.pad_size, self.pad_size, self.pad_size))  # self.pad_size = (self.window_size - 1) // 2
    q, k, v = self.qkv(x)
    attn = na2d_qk(q, k, rpb=None, kernel_size=self.kernel_size)  # batch,heads,H,W,self.window_size**2
    attn += self.rpb    # self.rpb =nn.Parameter(1,heads,1,1,self.window_size**2)
    attn += self.mask   # same as swin transformer
    attn = attn.softmax(dim=-1)
    x = na2d_av(attn, v, kernel_size=self.kernel_size)
    x=x[:,:,self.pad_size:-self.pad_size,self.pad_size:-self.pad_size,:]
```
like this, after padding, RPB is  simply  added to attn, and then a mask is applied to set the attention outside the image to 0.
In this case, would it be possible to add RPB without significantly increasing the computational cost?

**Comment by alihassanijr** [comment]
@Temo233233 yes that's correct. The inefficiency is not a matter of computation (FLOPs), but rather algorithmic.

I should clarify that the comment I made about RPB, or generally attention biases being inefficient, applies to **fused attention algorithms**, rather than unfused, which is what you're using above.

The inefficiency doesn't stem from NA allowing non-centered queries, but more generally applies to all fused attention kernels.

Fused attention kernels are complicated, and tend to even oversubscribe hardware resources like registers without any extra features like custom masks or attention bias. Adding anything significant to a fused attention kernel that achieves state of the art % of peak FLOPS is almost guaranteed to hurt performance.

While RPB seems like just an elementwise op in the forward pass, it's a **reduction** in the backward pass, and the fused backward pass kernel is already much more complex than the forward pass.

Even if implemented, it will still be a worse choice in theory compared to something like RoPE which is elementwise in both forward and backward pass.

The reason I brought even self attention into this argument is that if the number of attention biases/mask size is actually expressed in O(n^2), we're basically defeating the purpose of a fused attention kernel, by re-introducing O(n^2) memory operations, therefore making it so much easier for the kernel to wind up bound by memory bandwidth instead of compute.

I hope this helps clarify what I meant -- which again doesn't apply to unfused attention -- but if it doesn't, please feel free to follow up.
